File: check-ebird.py
#!/usr/bin/env python3

import fileinput
import csv

from utils import get_observations, species_name


def main():
    inat_observations, _ = get_observations()
    with fileinput.input() as ebird_csv_file:
        ebird_observations = csv.DictReader(ebird_csv_file)
        ebird_taxons = {
            obs["Scientific Name"]
            .replace(" (Domestic type)", "")
            .replace(" (Feral Pigeon)", ""): obs["Common Name"]
            for obs in ebird_observations
        }
        inat_taxons = {
            species_name(taxon.name): taxon.preferred_common_name
            for taxon in inat_observations.research_grade_taxons
            if taxon.iconic_taxon_name == "Aves"
        }
        inat_needs_id_taxons = {
            species_name(taxon.name) for taxon in inat_observations.needs_id_taxons
        }

        for taxon in ebird_taxons.keys() - inat_taxons.keys():
            needs_id = taxon in inat_needs_id_taxons
            print(
                f"Missing in iNaturalist:{' [NEEDS-ID]' if needs_id else ''} {ebird_taxons[taxon]} ({taxon})"
            )
        for taxon in inat_taxons.keys() - ebird_taxons.keys():
            print(f"      Missing in Ebird: {inat_taxons[taxon]} ({taxon})")


if __name__ == "__main__":
    main()

File: scripts/check-ebird.ts
#!/usr/bin/env -S npx tsx

import { createReadStream } from "node:fs";
import { parse } from "csv-parse";
import Config from "../src/lib/config";
import { getObservations, speciesName } from "../src/lib/inat/observations";

const NAME_OVERRIDES: Record<string, string> = {
  "Larus occidentalis x glaucescens": "Larus glaucescens × occidentalis",
};

function normalizeEbirdName(name: string): string {
  const stripped = name
    .replace(" (Domestic type)", "")
    .replace(" (Feral Pigeon)", "");
  return NAME_OVERRIDES[stripped] ?? stripped;
}

async function readEbirdTaxons(
  csvPath: string | undefined,
): Promise<Map<string, string>> {
  const source = csvPath ? createReadStream(csvPath) : process.stdin;
  const parser = source.pipe(parse({ columns: true }));

  const taxons = new Map<string, string>();
  for await (const row of parser as AsyncIterable<Record<string, string>>) {
    taxons.set(normalizeEbirdName(row["Scientific Name"]), row["Common Name"]);
  }
  return taxons;
}

async function main() {
  const csvPath = process.argv[2];
  const [{ summary }, ebirdTaxons] = await Promise.all([
    getObservations(Config.inatUserId),
    readEbirdTaxons(csvPath),
  ]);

  const inatTaxons = new Map(
    summary.researchGradeTaxons
      .filter((t) => t.iconic_taxon_name === "Aves" && t.name)
      .map((t) => [speciesName(t.name!), t.preferred_common_name ?? ""]),
  );
  const inatNeedsIdTaxons = new Set(
    summary.needsIdTaxons
      .filter((t) => t.name)
      .map((t) => speciesName(t.name!)),
  );

  for (const [name, commonName] of ebirdTaxons) {
    if (inatTaxons.has(name)) continue;
    const needsId = inatNeedsIdTaxons.has(name) || name === "Larus livens";
    console.log(
      `Missing in iNaturalist:${needsId ? " [NEEDS-ID]" : ""} ${commonName} (${name})`,
    );
  }
  for (const [name, commonName] of inatTaxons) {
    if (ebirdTaxons.has(name)) continue;
    console.log(`      Missing in eBird: ${commonName} (${name})`);
  }
}

main();
